### DISCUSSION TITLE
How to install external C libraries?

### DISCUSSION THREAD

**Discussion by danielpox** [micropython/micropython#11519] [Libraries & Drivers]
I would like to use [Apple's HomeKit ADK](https://github.com/apple/HomeKitADK), written in C, from within MicroPython.

There is a [Python implementation](https://github.com/ikalchev/HAP-python) of the HomeKit Accessory Protocol, but I'm not sure if it works with MicroPython on the Raspberry Pi Pico W.

I realize that I would have to write my own custom user module, as a bridge between the actual HomeKit ADK and MicroPython.

But I have no idea how adding external libraries even work. I have added the HomeKit ADK as a submodule inside the /lib directory, but that's as far as I've got.

Do I have to compile the external library separately, like adding a custom Makefile making it work with MicroPython? Or does the MicroPython compilation phase figure it all out using the library's source files?

What's the next step in creating the "bridge" user module? How do I import the external library in my custom user module? Is it as simple as this?:
```c
#include "lib/HomeKitADK/{what file do I include?}"

```

How do I even know if this library will work on the port I want to use? (rp2, Pico W)
(There is a port of the HomeKit Accessory Protocol for the [ESP32](https://github.com/espressif/esp-apple-homekit-adk) and [ESP8266](https://github.com/Mixiaoxiao/Arduino-HomeKit-ESP8266).)

I'm completely lost. Can anyone guide me in the right direction?

Note: I have virtually no experience in writing C, but I guess I'll figure it out bit by bit. 😅

**Comment by jimmo** [comment]
In theory this is possible, but this is not going to be at all straightforward.

Here's how I would maybe approach it.

1. Learn how to use the Homekit ADK on its own. e.g. get it working with a demo program on Linux. Write your own device.
2. Write a simple "User C Module" for MicroPython that doesn't have any dependencies. Learn about how to write Python bindings.
3. Figure out how to add the ADK to the MicroPython build.
4. Figure out what the Python bindings need to look like (i.e. write a stub version in Python first).
5. Implement those bindings in C.

> Note: I have virtually no experience in writing C, but I guess I'll figure it out bit by bit. sweat_smile

I mean this with kindness not criticism... this might not be the best thing to learn C with... there are a _lot_ of parts here. Adding third-party code to a different existing build system is pretty much one of the most difficult things you can do in C.

I think probably your best option to figure out how to make the Python implementation you linked to work on MicroPython. That is also easier said than done, but still much more clear how to proceed.
